### ISSUE TITLE
non-NPCs should not have NPC style on Display Name in chat, Hall, etc

### ISSUE THREAD

**Issue by Alys** [HabitRPG/habitica#11796]
Many players now have the NPC style on their Display Names, even though they are not NPCs. For example:
- lefnire
- Lemoness
- Megan

I believe it's happening to anyone who supported Habitica's original Kickstarter and is due to the code below, which sets `isNPC` to true if the user's account has any value for `user.backer.level`
https://github.com/HabitRPG/habitica/blob/257a52cba1712f4fbc5b0265fd03dbd4fe1a1afe/website/client/src/components/userLink.vue#L93-L100

A backer is an NPC only if they have user.backer.level set to 800. For more details, see the top post in https://github.com/HabitRPG/habitica/issues/11617 which shows the precise data that exists in the account of an NPC.

I'm setting this to high priority because the NPC title was earned by donating $800 in Habitica's and anyone who didn't donate that much shouldn't look like they did.

Anyone who states that they want to work on this should be willing to start immediately and submit a pull request rapidly. 

---

The NPC label was also applied to my account because I was a backer but I've temporarily deleted my backer details to fix that for me. I'm not willing to do that for anyone else of course. For my reference, these were my backer details: `"backer": { "tokensApplied": true, "tier": 80 },`

Thanks to cTheDragons and JohnJSal for noticing and reporting this! Much appreciated.

**Comment by Alys** [comment]
Anyone who states that they want to work on this should be willing to start immediately and submit a pull request rapidly.

**Comment by nikonhub** [comment]
Hello @Alys 

I can hix this one.

**Comment by Alys** [comment]
@Nikosmonaut Awesome, thanks so much! Go right ahead!

**Comment by nikonhub** [comment]
I need a little help here :).
Because depending on different functions there is a different comparison.

Those are 3 that I found : 

- 
```
 // UserLink
 if (this.backer) {
        return this.backer.tier;
      } if (this.user && this.user.backer) {
        return this.user.backer.tier;
      }
```

- Not sure about this one. What is "npc" field, should I use it ?
```
// AutoComplete
const isNPC = Boolean(message.backer && message.backer.npc);
```

- This one is simple. If I look at the user mongodb schema there is no "level". I suppose it is "tier" field now.
```
      // UserLabel
      if (this.backer) {
        return this.backer.level;
      } if (this.user && this.user.backer) {
        return this.user.backer.level;
      }
```

This is mongodb user schema :
```
  backer: {
    tier: Number,
    npc: String,
    tokensApplied: Boolean,
  },
```

**Comment by Alys** [comment]
@Nikosmonaut Ah yeah, some of that code is just wrong. It should be `backer.tier` never `backer.level`, exactly as you say. Sorry, I missed that when I was writing this issue.

Using `message.backer.tier === 800` should give identical results to using `message.backer.npc` since the `backer.npc` title was given to all players who had 800 for their `backer.tier`. The tier amount is the primary data that defines whether someone is an NPC.

Thanks for the pull request! I'll try to make time to test it later today.

**Comment by Alys** [comment]
I wanted to do some thorough testing for the fix for this issue so I compiled a list of all the places that a user's names appear. It kind of got bigger than I initially intended but I decided I might as well be thorough while I was doing this. :laughing: Several of the locations below are not actually relevant for this issue.

All of these locations have both Username and Display Name unless stated otherwise.

_For the "BTW" comments, ignore them for now; we can create issues later if desired._

- in website header stats area
- in website header on-hover stats area for Party mates
- at top of profile screen (to the right of the avatar)
- at top of profile tab on profile screen (under the avatar)
- at top of User Icon drop-down menu (Display Name only) (we don't want contributor tier/NPC styles for this one; I'm just including it for completeness)
- in left-hand panel on private messages page (index of all conversations)
- in your own messages in the right-hand panel on private messages page
- in other player's messages in the right-hand panel on private messages page
- at top of private messages page when one conversation is selected (Display Name only) _(BTW this should probably be changed to Username or to both names)_
- in right-hand panel on Tavern page: list of staff and mods, including Bailey who is both mod and NPC (Display Name only) _(BTW this should probably be changed to Username or to both names)_
- in right-hand panel on Tavern page: list of Player Tiers (no player names but contributor tier colours and icons appear, including for NPCs)
- in messages in Tavern / Guild / Party chats (NB player's title appears on hover but it's not styled, which is correct)
- in "@usernames" mentions in chat posts (we don't want contributor tier/NPC styles for this one; I'm just including it for completeness)
- in the "Group Leader" field at the top of a Guild or Party (Display Name only) _(BTW this should probably be changed to Username or to both names)_
- in the Members list for a Guild or Party
- in the Invites list for a Guild or Party
- in the "Created By" field on the "My Challenges" and "Discover Challenges" index pages (Display Name only) _(BTW this should probably be changed to Username, or perhaps to both names if they can be made to fit)_
- in the "Created By" field on an individual Challenge's page (Display Name only) _(BTW this should probably be changed to Username or to both names)_
- in the Participants list for a Challenge
- in the "To ... From ..." section on a greeting card modal (e.g., "To: Alys, From: A Little Yellow Spider") (Display Name only for both users) _(BTW this should probably be changed to Username or to both names for both users)_
- in `system` messages in Party chat (quest progress, transformation buffs, skill casts, and maybe Group Plan task messages) (we don't want contributor tier/NPC styles for this one; I'm just including it for completeness) (Display Name only for all of them I think) _(BTW this should probably be changed to Username)_
- in the Participants list for an active quest
- in the Participants list for a quest in the invitation stage (Display Name only) _(BTW this should probably be changed to Username or perhaps to both names if they can be made to fit)_
- in contributor credit messages in the "New Stuff by Bailey" modals (we don't want contributor tier/NPC styles for this one; I'm just including it for completeness)
- in the "Assigned To" button in the task edit screen for Group Plan tasks (we don't want contributor tier/NPC styles for this one; I'm just including it for completeness) (Display Name only) _(BTW this should probably be changed to Username)_
- in the "Assigned To" modal that you get to from the task edit screen for Group Plan tasks (we might not want contributor tier/NPC styles for this one) (Display Name only) _(BTW this should probably be changed to Username or perhaps to both names if they can be made to fit)_
- _(BTW, this is almost totally unrelated but while I'm thinking of it... at least some emails that reference other players have Display Names not Usernames; perhaps that should be changed?)_
- **Anywhere else on the website? If anyone finds any locations not included here, please add a comment to this issue, even if this issue has been closed because we'll refer back to this list later and/or copy it to an appropriate wiki page.**

**Comment by Alys** [comment]
Breaking up that long list into two sections...

On the production site (habitica.com), the following items DON'T display any special formatting, so when testing the fix for this issue, I'll check that these items either still DON'T display formatting or DO display formatting but appropriately. If new formatting has appeared and I'm in any doubt about whether that's good, I'll ping the staff about it.

- in website header stats area
- in website header on-hover stats area for Party mates
- at top of profile screen (to the right of the avatar)
- at top of profile tab on profile screen (under the avatar)
- at top of User Icon drop-down menu (Display Name only) (we don't want contributor tier/NPC styles for this one)
- in "@usernames" mentions in chat posts (we don't want contributor tier/NPC styles for this one)
- in the "Group Leader" field at the top of a Guild or Party (Display Name only)
- in the Members list for a Guild or Party ([admins can test in this private guild](https://habitica.com/groups/guild/0b306146-e281-4f85-8a58-b05cddcaf219))
- in the Invites list for a Guild or Party
- in the "Created By" field on an individual Challenge's page (Display Name only)
- in the Participants list for a Challenge
- in the "To ... From ..." section on a greeting card modal (e.g., "To: Alys, From: A Little Yellow Spider") (Display Name only for both users)
- in `system` messages in Party chat (quest progress, transformation buffs, skill casts, and maybe Group Plan task messages) (we don't want contributor tier/NPC styles for this one) (Display Name only for all of them I think)
- in the Participants list for an active quest
- in the Participants list for a quest in the invitation stage (Display Name only)
- in contributor credit messages in the "New Stuff by Bailey" modals (we don't want contributor tier/NPC styles for this one)
- in the "Assigned To" button in the task edit screen for Group Plan tasks (we don't want contributor tier/NPC styles for this one) (Display Name only)
- in the "Assigned To" modal that you get to from the task edit screen for Group Plan tasks (we might not want contributor tier/NPC styles for this one) (Display Name only)

On the production site, the following items DO have the contributor tier/NPC formatting, so when testing the fix for this issue, I'll check that these items use the correct formatting:

- in left-hand panel on private messages page (index of all conversations)
- in your own messages in the right-hand panel on private messages page
- in other player's messages in the right-hand panel on private messages page
- at top of private messages page when one conversation is selected (Display Name only)
- in right-hand panel on Tavern page: list of staff and mods, including Bailey who is both mod and NPC (Display Name only)
- in right-hand panel on Tavern page: list of Player Tiers (no player names but contributor tier colours and icons appear, including for NPCs)
- in messages in Tavern / Guild / Party chats (player's title on hover should not be styled)
- in the "Created By" field on the "My Challenges" and "Discover Challenges" index pages (Display Name only)

I'll put my test results in the PR. I'm just putting this list here now in case we need to use it later for other fixes.

**Comment by Alys** [comment]
Ha. :) Those two lists of locations need to be adjusted already now that https://github.com/HabitRPG/habitica/pull/11763 has been merged. :laughing: Some locations that used to not have formatting now do have it. This is just a reminder to me that I'll want to re-sort the lists again the next time I need to test name formatting.

**Comment by citrusella** [comment]
Might-be-rather-important-but-IDK-if-anyone-will-want-to-do-anything-about-it-if-the-fix-itself-doesn't-resolve-it note: The NPC formatting appears to apply retroactively to *all posts* of affected users. My reference points for this are:

- posts from several months back from Lemoness in the [Seasonal Decoration Guild](https://habitica.com/groups/guild/f065a669-9483-42d4-9562-7f8dd5b8adb7) (I was looking at it because I know she frequents it sometimes and thought it might let me see the bug in action)
- my PM history with @Alys (the messages themselves mark her as NPC, though the sidebar matches her current displayed tier of mod)

From the second bullet I think it could be assumed (or at least expected/looked for? IDK, maybe whatever fix will retroactively fix the posts too) that the name might continue to display wrong even after the issue is fixed, if the fact that Alys' restoring of her modness by removing her backer tier (but that not applying to fixing her prior posts) is any indication.

I've made an educated guess this is related to the issue because my own name in a chat old enough to show me pre-tier-7 still has my old tier formatting applied so it's NPC-specific and not just anyone with a tier having it go retroactively.

For all I know this will be totally and completely fixed by fixing the issue (like, it'll suck the NPC color away where it shouldn't be but leave the intended color or something), but I figured it was better to bring it up so it's known just in case it's still an issue after.

**Comment by Alys** [comment]
@citrusella Thank you for reporting that! The fix should apply retroactively as you suggest because it will be applying a style to the name based on backer information stored in post. Fortunately this bug doesn't affect the data stored in the post, just how the styles are chosen based on that data.

I've realised something in my top post in this issue was misleading: "I've temporarily deleted my backer details to fix that for me." -- that applies only to posts that I made after I deleted my backer data. My older posts still have my backer data stored in them, so they'll still have the NPC style at the moment.
